### ISSUE TITLE
PHP 8 support

### ISSUE THREAD

**Issue by peterfabian** [woocommerce/woocommerce#27674]
Similarly to WordPress 5.6 (planned for 8 December 2020), we need to start preparations to be fully compatible with PHP 8 once it lands on November 26, 2020.

We need to
- make PHPUnit compatible, probably based on the approach WP core is taking: https://core.trac.wordpress.org/ticket/46149
- examine the changes in https://github.com/php/php-src/blob/master/UPGRADING#L20 and perhaps [other guides](https://href.li/?https://stitcher.io/blog/new-in-php-8#breaking-changes) to make sure we won't miss anything

WP core trac ticket for some reference: https://core.trac.wordpress.org/ticket/50913

Related: p59oML-1O4-p2

**Related pull requests:**

- [x] https://github.com/woocommerce/woocommerce/pull/27830 - Add a custom round method to workaround an incompatibility of the buil-in function
- [x] https://github.com/woocommerce/woocommerce/pull/27840 - Convert optional-before-required parameters into required parameters
- [x] https://github.com/woocommerce/woocommerce/pull/27844 - Use a custom version of PHPUnit 7 (temporary workaround for now) - [the customized PHPUnit](https://github.com/woocommerce/phpunit/pull/1)
- [x] https://github.com/woocommerce/woocommerce/pull/27848 - Remove a useless test assertion that fails in PHP 8
- [x] https://github.com/woocommerce/woocommerce/pull/27707 - Change __wakeup() to a public method
- [x] https://github.com/woocommerce/woocommerce/pull/27986 - Fixes in the dependency injection engine for PHP 8 compatibility
- [x] https://github.com/woocommerce/woocommerce/pull/27987 - Convert optional-before-required parameters into required parameters (take 2)

Checked items are already merged.

**Comment by Preciousomonze** [comment]
Wow, this is fast, good one people 🤯🥳

**Comment by peterfabian** [comment]
Milestoning for 4.7, as that is the last release before WP 5.6 gets released, but not everything might be ready by then, so we might need to adjust the plan.

**Comment by peterfabian** [comment]
Errors seen in initial testing: 
FireShot Capture 047 -  - woo.loc
FireShot Capture 050 - Profiler ‹ WooCommerce ‹ wootest — WooCommerce - woo.loc

**Comment by Konamiman** [comment]
I've found that when trying to manually create an order, `WC_Abstract_Order->get_total_discount` does a `round('')` which results in `CRITICAL Uncaught TypeError: round(): Argument #1 ($number) must be of type int|float, string given`

**Comment by peterfabian** [comment]
Latest update from WP core: 

> Sergeybiryukov shared that work on PHP 8 support has continued. All known unit test issues have been addressed, and the tests are now passing on PHP 8. #50913 includes most of the progress on this, some work was also done in other related tickets: https://core.trac.wordpress.org/query?keywords=~php8

> Next steps for PHP 8 support:

> Docker image needs to be updated to use the same PHP extensions for PHP 8 that we have on PHP 7.x: gd, mbstring, etc. There is a PR by Desrosj waiting for review: https://github.com/WordPress/wpdev-docker-images/pull/36. If anyone is able to help with that, please do.
There is an ongoing effort to fix some function argument type issues reported by PHPStan (a static analyzer): #51423.
More testing on PHP 8, expanding test coverage, and creating tickets for any issues found. Andraganescu and Desrosj are working on a call for testing to be published later.

**Comment by peterfabian** [comment]
Updates from WP core: https://make.wordpress.org/core/2020/10/06/call-for-testing-php-8-0/

**Comment by claudiosanches** [comment]
@peterfabian Currently it's missing only to update our Unit Tests, but this isn't something that depends on a release and can be merged at any time. I'm going to remove the 4.7 milestone, since 4.7 already includes all other changes.

@Konamiman Thank you very much for all your work!

**Comment by peterfabian** [comment]
Just trying to add PHP 8 to our CI as part of https://github.com/woocommerce/woocommerce/pull/27844 and we can close this one.

**Comment by Konamiman** [comment]
Closing as we've recently merged https://github.com/woocommerce/woocommerce/pull/27844, which was the last pending PR on this subject.

**Comment by genuino2** [comment]
FATAL ERROR with php 8 on **customer-processing-order.php** on:
`do_action( 'woocommerce_email_order_meta', $order, $sent_to_admin, $plain_text, $email );
`

**Comment by Konamiman** [comment]
Hi @genuino2, thanks for your report. Could you please provide more details on what the exact error is and what are the steps to get it?

**Comment by genuino2** [comment]
Sure: please note that the custom template used is different only for one
paragraph. I've tried also with original template and the error was the
same.
Please, pay attention to row that using my template. Is a little different
from the original file.


[25-Feb-2022 14:47:52 UTC] PHP Fatal error:  Uncaught TypeError: Illegal
offset type in isset or empty in
/home2/lwgaspar/public_html/wp-includes/meta.php:657
Stack trace:
#0 /home2/website/public_html/wp-includes/meta.php(572):
get_metadata_raw('post', 155188, Array, true)
#1 /home2/website/public_html/wp-includes/post.php(2505):
get_metadata('post', 155188, Array, true)
#2
/home2/website/public_html/wp-content/plugins/woocommerce/includes/class-wc-emails.php(505):
get_post_meta(155188, Array, true)
#3 /home2/website/public_html/wp-includes/class-wp-hook.php(309):
WC_Emails->order_meta(Object(Automattic\WooCommerce\Admin\Overrides\Order),
false, false)
#4 /home2/website/public_html/wp-includes/class-wp-hook.php(331):
WP_Hook->apply_filters('', Array)
#5 /home2/website/public_html/wp-includes/plugin.php(474):
WP_Hook->do_action(Array)
#6
/home2/website/public_html/wp-content/themes/mytheme-child/woocommerce/emails/customer-processing-order.php(71):
do_action('woocommerce_ema...',
Object(Automattic\WooCommerce\Admin\Overrides\Order), false, false,
Object(WC_Email_Customer_Processing_Order))
#7
/home2/website/public_html/wp-content/plugins/woocommerce/includes/wc-core-functions.php(345):
include('/home2/website...')
#8
/home2/website/public_html/wp-content/plugins/woocommerce/includes/wc-core-functions.php(364):
wc_get_template('emails/customer...', Array, '', '')
#9
/home2/website/public_html/wp-content/plugins/woocommerce/includes/emails/class-wc-email-customer-processing-order.php(113):
wc_get_template_html('emails/customer...', Array)
#10
/home2/website/public_html/wp-content/plugins/woocommerce/includes/emails/class-wc-email.php(541):
WC_Email_Customer_Processing_Order->get_content_html()
#11
/home2/website/public_html/wp-content/plugins/woocommerce/includes/emails/class-wc-email-customer-processing-order.php(93):
WC_Email->get_content()
#12 /home2/website/public_html/wp-includes/class-wp-hook.php(307):
WC_Email_Customer_Processing_Order->trigger(155188,
Object(Automattic\WooCommerce\Admin\Overrides\Order))
#13 /home2/website/public_html/wp-includes/class-wp-hook.php(331):
WP_Hook->apply_filters('', Array)
#14 /home2/website/public_html/wp-includes/plugin.php(522):
WP_Hook->do_action(Array)
#15
/home2/website/public_html/wp-content/plugins/woocommerce/includes/class-wc-emails.php(170):
do_action_ref_array('woocommerce_ord...', Array)
#16 /home2/website/public_html/wp-includes/class-wp-hook.php(307):
WC_Emails::send_transactional_email(Array,
Object(Automattic\WooCommerce\Admin\Overrides\Order))
#17 /home2/website/public_html/wp-includes/class-wp-hook.php(331):
WP_Hook->apply_filters('', Array)
#18 /home2/website/public_html/wp-includes/plugin.php(474):
WP_Hook->do_action(Array)
#19
/home2/website/public_html/wp-content/plugins/woocommerce/includes/class-wc-order.php(373):
do_action('woocommerce_ord...', 155188,
Object(Automattic\WooCommerce\Admin\Overrides\Order))
#20
/home2/website/public_html/wp-content/plugins/woocommerce/includes/class-wc-order.php(222):
WC_Order->status_transition()
#21
/home2/website/public_html/wp-content/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-data.php(635):
WC_Order->save()
#22 /home2/website/public_html/wp-includes/class-wp-hook.php(309):
WC_Meta_Box_Order_Data::save(155188)
#23 /home2/website/public_html/wp-includes/class-wp-hook.php(331):
WP_Hook->apply_filters(NULL, Array)
#24 /home2/website/public_html/wp-includes/plugin.php(474):
WP_Hook->do_action(Array)
#25
/home2/website/public_html/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-meta-boxes.php(222):
do_action('woocommerce_pro...', 155188, Object(WP_Post))
#26 /home2/website/public_html/wp-includes/class-wp-hook.php(309):
WC_Admin_Meta_Boxes->save_meta_boxes(155188, Object(WP_Post))
#27 /home2/website/public_html/wp-includes/class-wp-hook.php(331):
WP_Hook->apply_filters(155188, Array)
#28 /home2/website/public_html/wp-includes/plugin.php(474):
WP_Hook->do_action(Array)
#29 /home2/website/public_html/wp-includes/post.php(4687):
do_action('save_post', 155188, Object(WP_Post), true)
#30 /home2/website/public_html/wp-includes/post.php(4789):
wp_insert_post(Array, false, true)
#31 /home2/website/public_html/wp-admin/includes/post.php(426):
wp_update_post(Array)
#32 /home2/website/public_html/wp-admin/post.php(227): edit_post()
#33 {main}
  thrown in /home2/website/public_html/wp-includes/meta.php on line 657

Il giorno ven 25 feb 2022 alle ore 16:53 Néstor Soriano <
***@***.***> ha scritto:

**Comment by Konamiman** [comment]
Thanks @genuino2 but I'd need some more details:

* What WooCommerce version are you using?
* Which is the original theme in which you are also getting the error?
* Could you please provide the detailed usage steps that trigger the error?

Thank you!

**Comment by genuino2** [comment]
WooCommerce 6.2.1
Theme: blocksy
PHP: 8
What i've done: I've moved an order from On Hold to Processing and i've the
error. Email to customer not sent. Switched to Php 7.4 and it's all ok.

Il giorno lun 28 feb 2022 alle ore 15:43 Néstor Soriano <
***@***.***> ha scritto:

**Comment by dadart123** [comment]
Woocommerce 6.9.0
Hello everyone,
My site has this error if I use php 8.0 and later
1.	PHP Fatal error:  Uncaught TypeError: array_filter(): Argument #1 ($array) must be of type array, bool given in /wp-content/plugins/woocommerce/includes/class-wc-frontend-scripts.php:65
If I go back to php 7.4 the problem is not there
Can you help me solve the problem in order to upgrade to PHP 8.00
Thank you very much
Eleonora

 Fixed with this solution
Please add this to your child theme's functions.php to temporarily resolve this issue:

add_filter( 'woocommerce_enqueue_styles', 'wc_child_theme_fix_33467', PHP_INT_MAX );

function wc_child_theme_fix_33467( $styles ) {
    return ( is_array( $styles ) ) ? $styles : [];
}

**Comment by beaulebens** [comment]
@dadart123 this should be addressed in 6.9.1, per https://github.com/woocommerce/woocommerce/pull/34671

**Comment by gg4midas** [comment]
woocommerce 7.1
theme: astra
My site has this error if I use PHP 8.1 and 8.0:

Error Details
=============
An error of type E_ERROR was caused in line 37 of the file /www/wwwroot/sweetener/wp-content/plugins/woocommerce/includes/class-wc-regenerate-images-request.php. Error message: Uncaught Error: Call to undefined function putenv() in /www/wwwroot/sweetener/wp-content/plugins/woocommerce/includes/class-wc-regenerate-images-request.php:37
Stack trace:
#0 /www/wwwroot/sweetener/wp-content/plugins/woocommerce/includes/class-wc-regenerate-images.php(49): WC_Regenerate_Images_Request->__construct()
#1 /www/wwwroot/sweetener/wp-includes/class-wp-hook.php(308): WC_Regenerate_Images::init()
#2 /www/wwwroot/sweetener/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters()
#3 /www/wwwroot/sweetener/wp-includes/plugin.php(517): WP_Hook->do_action()
#4 /www/wwwroot/sweetener/wp-settings.php(617): do_action()
#5 /www/wwwroot/sweetener/wp-config.php(100): require_once('...')
#6 /www/wwwroot/sweetener/wp-load.php(50): require_once('...')
#7 /www/wwwroot/sweetener/wp-admin/admin.php(34): require_once('...')
#8 /www/wwwroot/sweetener/wp-admin/index.php(10): require_once('...')
#9 {main}
thrown

**Comment by beaulebens** [comment]
@gg4midas please follow #31942 for discussion of the specific issue you're seeing.

**Comment by michaelbgibs** [comment]
I keep getting this error message and it's a core alert.  I have no idea why and it only happens when I use any php version higher than 7.4 with WordPress 6.22.

Uncaught Error: array_filter(): Argument #1 ($array) must be of type array, null given

wp-includes/class-wp-user-query.php:479


Any when I try to edit any page, I get a Fatal response:


array_filter() expects parameter 1 to be array, null given

wp-includes/class-wp-user-query.php:479
array_filter()
wp-includes/class-wp-user-query.php:479
WP_User_Query->prepare_query()
wp-includes/class-wp-user-query.php:78
WP_User_Query->__construct()
wp-includes/user.php:775
get_users()
wp-includes/user.php:1667
wp_dropdown_users()
wp-admin/includes/class-wp-posts-list-table.php:1711
WP_Posts_List_Table->inline_edit()
wp-admin/edit.php:493

Can you help me?

**Comment by Konamiman** [comment]
Hi @michaelbgibs. Could you please confirm if this error appears only if WooCommerce is installed? If that's the case, could you please create [a new issue](https://github.com/woocommerce/woocommerce/issues/new/choose) with detailed reproduction steps? Thank you!
